magma/backend/mlir/hardware_module.py

```diff
@@ -143,7 +143,7 @@ def magma_type_to_mlir_type(type: Kind) -> MlirType:
     if issubclass(type, m_Tuple):
         fields = (
             (magma_tuple_key_to_str(k, type), magma_type_to_mlir_type(t))
-            for k, t in type.field_dict.items()
+            for k, t in type.sorted_field_dict_items()
```

Comment:
I was running into an issue where struct fields were being generated in a different order in different places in the generated MLIR.  Perhaps there some places where we're transferring the data through some structure that changes the order or something, I can see if I can trace down the root cause location.

Reply:
So I think the issue is not that we create any incompatible orderings internally,  but that the user could produce wireable types with an unordered data structure, and since we do structural wiring, this is okay.  Here's some example code that works
```
x = {"x": 0, "y": 1}
y = {"y": 0, "x": 1}

assert list(x.keys()) != list(y.keys())
assert x.keys() == y.keys()


import magma as m

T0 = m.AnonProduct[{"x": m.Bit, "y": m.Bits[8]}]
T1 = m.AnonProduct[{"y": m.Bits[8], "x": m.Bit}]


class Foo(m.Circuit):
    io = m.IO(I=m.In(T0), O=m.Out(T0))
    io.O @= io.I


m.compile("build/Foo", Foo, output="mlir-verilog")
```

Now MLIR/CIRCT on the other hand, requires the tuple fields to be consistent in order across the IR code, so while in magma we may wire types with different orderings, we need to sort them for MLIR.

I will add a test case that covers this.

Reply:
Got it, thanks for hunting this down.

I feel like there should be something more explicit here. Ideally what is allowed in magma is consistent with what is allowed by the backend.

Here are some options (all of which btw, I think should be implemented in a separate PR fixing this):

1. Disallow magma from wiring tuples with keys in different orders. So in the above case `io.O @= io.I` would itself raise an error -- the same way that `m.Bit() @= m.Bits[8]()` would even raise an error -- just two different types that can't be wired.
2. All tuple keys should be inherently sorted. So `T0` and `T1` are the *same* type inherently (but multiple different constructor calls could lead to the same type). That way MLIR is agnostic to this. Right now the backend *silently* re-orders/sorts the keys. The keys should either always be considered ordered or not -- not half and half.

Don't have a strong opinion on this. (2) seems easier but it will likely break a lot of things, since it will re-order all keys. (1) seems somewhat difficult to detect...

Reply:
See #1314

Reply:
On one hand, OCAML does not care about the order of a record type (https://cs3110.github.io/textbook/chapters/data/records_tuples.html#:~:text=The%20order%20of%20the%20fi%3Ati%20inside%20a%20record%20type,could%20be%20used%20without%20definition.)

On the other hand, C does (since it allows you to control the layout).

I think 2. would actually provide better compatibility since existing code would still work (but the ordering in the backend might be different).  The tests here seem to suggest that's not a major issue for our internal test cases at least (we'd just be changing "when" the sorting happens to construction time rather than in the backend).  I think this option would mirror ocaml (there's no difference between orderings, and they are "normalized" via sorting the keys).

1. would provide more explicit control over layout (e.g. if we want fields to be flattened into specific "index"), which could be useful for performance control.

A middle ground might be to provide 1. as a low-level interface and define a more abstract type system (2.) on top of 1.